src/threshold-ts/staking/index.ts

```diff
@@ -249,21 +266,86 @@ export class Staking implements IStaking {
   }
 
   getOwnerStakes = async (owner: string): Promise<Array<Stake>> => {
-    const stakedEvents = (
+    const stakes = (
       await getContractPastEvents(this._staking, {
         eventName: "Staked",
         fromBlock: this.STAKING_CONTRACT_DEPLOYMENT_BLOCK,
         filterParams: [undefined, owner],
       })
-    ).reverse()
+    )
+      .map((event) => ({
+        stakingProvider: event.args?.stakingProvider,
+        stakeType: event.args?.stakeType,
+      }))
+      .reverse()
+
+    const { current, outdated } = await this.findRefreshedKeepStakes(owner)
+
+    const stakingProviders = stakes
+      .filter(({ stakingProvider }) => !outdated.includes(stakingProvider))
+      .concat(
+        current.map((stakingProvider) => ({
+          stakingProvider,
+          stakeType: StakeType.KEEP,
+        }))
+      )
 
     return await Promise.all(
-      stakedEvents.map((stakedEvent) =>
-        this.getStakeByStakingProvider(
-          stakedEvent.args?.stakingProvider,
-          stakedEvent.args?.stakeType
-        )
+      stakingProviders.map(({ stakingProvider, stakeType }) =>
+        this.getStakeByStakingProvider(stakingProvider, stakeType)
       )
     )
   }
+
+  findRefreshedKeepStakes = async (
+    owner: string
+  ): Promise<OwnerRefreshedResult> => {
+    // Find all events where the `owner` was set as a new owner or old owner of
+    // the stake.
+    const ownerRefreshedEvents = await getContractPastEvents(this._staking, {
+      eventName: "OwnerRefreshed",
+      filterParams: [null, owner, owner],
+      fromBlock: this.STAKING_CONTRACT_DEPLOYMENT_BLOCK,
+    })
+
+    // Convert to `Set` to remove duplicated staking provider addresses(the
+    // owner can be changed multiple times for the same staking provider) and
+    // then to an array so we can use `map`, `filter` and other array methods.
+    const possibleStakingProviders: string[] = Array.from(
+      new Set(ownerRefreshedEvents.map((event) => event?.args?.stakingProvider))
+    )
+
+    const multicalls: ContractCall[] = possibleStakingProviders.map(
+      (stakingProvider) => ({
+        address: this._staking.address,
+        interface: this._staking.interface,
+        method: "rolesOf",
+        args: [stakingProvider],
+      })
+    )
+
+    // For each staking provider, we need to verify that the current owner of
+    // the stake is indeed the one we are looking for(the `owner` argumment).
+    // It's possible that the owner can be changed multiple times. Instead of
+    // iterating through the `OwnerRefreshed` events and comparing the
+    // `oldOwner` and `newOner` params from that event, we can just check the
+    // current owner for a given staking provider by calling `rolesOf`.
+    const rolesOf = await this._multicall.aggregate(multicalls)
+
+    // Staking provideres are owned by the `owner` address.
```

Comment:
I'm not sure but I believe this comment is describing what `stakingProviders` variable is. If so the I would change it to:

```suggestion
    // Staking provideres that are owned by the `owner` address.
```

Reply:
9539bfa